File: Farseer_Project/Farseer/Farseer.m
//
//  FSLog.m
//  Farseer
//
//  Created by Go Salo on 1/30/15.
//  Copyright (c) 2015 so.salo. All rights reserved.
//

#include "Farseer.h"
#include <stdio.h>
#include <stdarg.h>
#include <time.h>
#include <string.h>
#include <assert.h>
#include <stdlib.h>
#include <unistd.h>
#include <pwd.h>
#include <sys/types.h>
#include <dirent.h>
#import <Foundation/NSString.h>
#import "FSLogManager.h"
#import "FSDebugCentral.h"

static BOOL isDebug = NO;
static FSLogLevel consoleLevel = FSLogLevelError;

void FS_DebugLog(NSInteger code, NSString *domain, FSLogLevel level, NSDictionary <NSString *, id>*info, const char *file, const char *function, unsigned int line)
{
    FSBLELog *log = [FSBLELog createLogWithLevel:level domain:domain info:info file:file function:function line:line];
    [[FSDebugCentral getInstance].logManager inputLog:log];
    
    if (isDebug) {
        const char *cLog = [domain cStringUsingEncoding:NSUTF8StringEncoding];
        char *prefix = NULL;
        switch (level) {
            case FSLogLevelFatal:
                prefix = "[FATAL]";
                break;
            case FSLogLevelError:
                prefix = "[ERROR]";
                break;
            case FSLogLevelWarning:
                prefix = "[WARNING]";
                break;
            case FSLogLevelLog:
                prefix = "[LOG]";
                break;
            case FSLogLevelMinor:
                prefix = "[MINOR]";
                break;
            default:
                assert(false);
                return;
        }
        
        if (level >= consoleLevel)
        {
            printf("%s:%s\n", prefix, cLog);
        }
        
        if (level == FSLogLevelFatal)
        {
            printf("fatal error");
            assert(false);
        }
    }
    
}

void FSSendLog(id<FSBLELogProtocol, FSStorageLogProtocol> log) {
    [[FSDebugCentral getInstance].logManager inputLog:log];
}

void closeBLEDebug() {
    [FSDebugCentral closeBLEDebug];
}

void openBLEDebug(void(^callback)(NSError *error)) {
    [FSDebugCentral openBLEDebug:callback];
}

void cleanLogBefore(NSDate *date) {
    [[FSDebugCentral getInstance].logManager cleanLogBeforeDate:date];
}

File: Farseer_Project/Farseer/Farseer.swift
//
//  Farseer.swift
//  SLFarseer
//
//  Created by Go Salo on 15/3/31.
//  Copyright (c) 2015年 Qeekers. All rights reserved.
//

import Foundation

public func FSLog(log: String, fileName: String = __FILE__, functionName: String = __FUNCTION__, lineNumber: Int = __LINE__) {
    FS_DebugLog(log, .Log, fileName, functionName, UInt32(lineNumber))
}

public func FSWarning(log: String, fileName: String = __FILE__, functionName: String = __FUNCTION__, lineNumber: Int = __LINE__) {
    FS_DebugLog(log, .Warning, fileName, functionName, UInt32(lineNumber))
}

public func FSError(log: String, fileName: String = __FILE__, functionName: String = __FUNCTION__, lineNumber: Int = __LINE__) {
    FS_DebugLog(log, .Error, fileName, functionName, UInt32(lineNumber))
}

public func FSFatal(log: String, fileName: String = __FILE__, functionName: String = __FUNCTION__, lineNumber: Int = __LINE__) {
    FS_DebugLog(log, .Fatal, fileName, functionName, UInt32(lineNumber))
}

public func FSMinor(log: String, fileName: String = __FILE__, functionName: String = __FUNCTION__, lineNumber: Int = __LINE__) {
    FS_DebugLog(log, .Minor, fileName, functionName, UInt32(lineNumber))
}

public func FSCLog(condition: Bool, log: String, fileName: String = __FILE__, functionName: String = __FUNCTION__, lineNumber: Int = __LINE__) {
    if condition { FS_DebugLog(log, .Log, fileName, functionName, UInt32(lineNumber)) }
}

public func FSCWarning(condition: Bool, log: String, fileName: String = __FILE__, functionName: String = __FUNCTION__, lineNumber: Int = __LINE__) {
    if condition { FS_DebugLog(log, .Warning, fileName, functionName, UInt32(lineNumber)) }
}

public func FSCError(condition: Bool, log: String, fileName: String = __FILE__, functionName: String = __FUNCTION__, lineNumber: Int = __LINE__) {
    if condition { FS_DebugLog(log, .Error, fileName, functionName, UInt32(lineNumber)) }
}

public func FSCFatal(condition: Bool, log: String, fileName: String = __FILE__, functionName: String = __FUNCTION__, lineNumber: Int = __LINE__) {
    if condition { FS_DebugLog(log, .Fatal, fileName, functionName, UInt32(lineNumber)) }
}

public func FSCMinor(condition: Bool, log: String, fileName: String = __FILE__, functionName: String = __FUNCTION__, lineNumber: Int = __LINE__) {
    if condition { FS_DebugLog(log, .Minor, fileName, functionName, UInt32(lineNumber)) }
}